src/compiler/checker.ts

```diff
@@ -22316,6 +22406,20 @@ namespace ts {
                 return result;
             }
 
+            // At flow control branch or loop junctions, if the type along every antecedent code path
+            // is an evolving array type, we construct a combined evolving array type. Otherwise we
+            // finalize all evolving array types.
+            function getUnionOrEvolvingArrayType(types: Type[], subtypeReduction: UnionReduction) {

```

Comment:
Rather than closing over `declaredType`, could we not pass it in as an argument? That way `getUnionOrEvolvingArrayType` can stay outside of `isEvolvingArrayOperationTarget` so there is no need for a unique copy of the function to exist each time `isEvolvingArrayOperationTarget` is called (as each copy gets bound to the new local environment record to reference the closed-over variable).

Reply:
Sure, we could do that, although many other local functions inside the `getFlowTypeOfReference` function do the same.

Reply:
I only mention this one because this PR moves it from outside to inside so that it can observe the `declaredType` variable, which seems unnecessary in this case.

Reply:
Right, but that's generally the case with many of the local functions inside `getFlowTypeOfReference` so I think it's ok.